java/ql/src/semmle/code/java/security/JWT.qll

```diff
@@ -0,0 +1,119 @@
+/** Provides classes for working with JWT libraries. */
+
+import java
+private import semmle.code.java.dataflow.ExternalFlow
+private import semmle.code.java.dataflow.DataFlow
+
+/** A method access that assigns signing keys to a JWT parser. */
+class JwtParserWithInsecureParseSource extends DataFlow::Node {
+  JwtParserWithInsecureParseSource() {
+    exists(MethodAccess ma, Method m |
+      m.getDeclaringType().getASupertype*() instanceof TypeJwtParser or
+      m.getDeclaringType().getASupertype*() instanceof TypeJwtParserBuilder
+    |
+      this.asExpr() = ma and
+      ma.getMethod() = m and
+      m.hasName(["setSigningKey", "setSigningKeyResolver"])
+    )
+  }
+}
+
+/**
+ * The qualifier of an insecure parsing method.
+ * That is, either the qualifier of a call to the `parse(token)`,
+ * `parseClaimsJwt(token)` or `parsePlaintextJwt(token)` methods or
+ * the qualifier of a call to a `parse(token, handler)` method
+ * where the `handler` is considered insecure.
+ */
+class JwtParserWithInsecureParseSink extends DataFlow::Node {
+  MethodAccess insecureParseMa;
+
+  JwtParserWithInsecureParseSink() {
+    insecureParseMa.getQualifier() = this.asExpr() and
+    exists(Method m |
+      insecureParseMa.getMethod() = m and
+      m.getDeclaringType().getASupertype*() instanceof TypeJwtParser and
+      m.hasName(["parse", "parseClaimsJwt", "parsePlaintextJwt"]) and
+      (
+        m.getNumberOfParameters() = 1
+        or
+        isInsecureJwtHandler(insecureParseMa.getArgument(1))
+      )
+    )
+  }
+
+  /** Gets the method access that does the insecure parsing. */
+  MethodAccess getParseMethodAccess() { result = insecureParseMa }
+}
+
+/** A set of additional taint steps to consider when taint tracking JWT related data flows. */
+class JwtParserWithInsecureParseAdditionalTaintStep extends Unit {
+  predicate step(DataFlow::Node node1, DataFlow::Node node2) {
+    jwtParserStep(node1.asExpr(), node2.asExpr())
+  }
+}
```

Comment:
Is this meant to be an extension point for the query? If so, then `predicate step` should be `abstract`.

Reply:
It is, this was a mistake. Thanks for pointing this out, fixed in b586f3e.

Reply:
I'm not sure, but when would a user want an additional flow steps for this?
But it will also probably not hurt to have this even if it will be used rarely^^

Reply:
Likely never, but we trying to move towards a more uniform query skeleton that allows common extension points.